### PR TITLE
Add generic types for results and collections

### PR DESCRIPTION
**Pull request by rodrigopivi**
Now if you declare a variable like:
```js
const ob = DB.objects<IUserContext>("UserContext").sorted("token", true);
```

You can expect that ob's type is `IUserContext`

Also works for creation of objects:
```js
DB.create(DB.objects<IRole>("Role"), { xx: "type checked field" });
```

### CODE REVIEW

File: realm/index.d.ts [hunk 1]
```diff
@@ -107,24 +107,24 @@ declare namespace Realm {
          * @param  {any[]} ...arg
          * @returns Results
          */
-        filtered(query: string, ...arg: any[]): Results;
+        filtered(query: string, ...arg: any[]): Results<T & Object>;
```
Comment (DanielRosenwasser) [inline realm/index.d.ts]: `T & Object` doesn't actually do anything. The apparent type of `T` is `{}` which is effectively the same as `Object` - in short, you get all the same members of `Object` anyway. Furthermore, `number` and `string` are assignable to `Object` and `{}`, so it doesn't do a ton for you.

I think what you really want to do is constrain `T` to be of only an object type, which TypeScript 2.2 will have. [See here](https://github.com/Microsoft/TypeScript/issues/1809)

Until then, I think you'll just have to roll with `T`.

**Review by DanielRosenwasser — DISMISSED** [review]

File: realm/index.d.ts (hunk 1 above)
Comment (DanielRosenwasser) [inline realm/index.d.ts]: Ignore the above - I just realized you were referring to `Realm.Object`. Can you please make that explicit?

**Comment by DanielRosenwasser** [comment]
Can you update the version number at the top of the file? Keep it as the MAJOR.MINOR version (and remove anything else following that)

Can you also add a `tslint.json` file that contains the following line?

```json
{ "extends": "../tslint.json" }
```

@Akim95 can you review this?

File: realm/index.d.ts (hunk 1 above)
Comment (rodrigopivi) [inline realm/index.d.ts]: Thanks for the review. Done, updated to make explicit.

**Comment by rodrigopivi** [comment]
@DanielRosenwasser Thanks for the review, done.

File: realm/index.d.ts [hunk 2]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for realm-js 0.14.3
+// Type definitions for realm-js 0.14.4
```
Comment (Akim95) [inline realm/index.d.ts]: Just change to realm-js 0.14

File: realm/index.d.ts [hunk 3]
```diff
@@ -1,4 +1,4 @@
-// Type definitions for realm-js 0.14.3
+// Type definitions for realm-js 1.0
```
Comment (Akim95) [inline realm/index.d.ts]: Please update version to realm-js 0.14

**Review by Akim95 — CHANGES_REQUESTED** [review]
Please update version to realm-js 0.14

File: realm/index.d.ts [hunk 4]
```diff
@@ -150,65 +150,65 @@ declare namespace Realm {
         /**
          * @param  {number} start?
          * @param  {number} end?
-         * @returns Object
+         * @returns T[] | Object[]
          */
-        slice(start?: number, end?: number): Object[];
+        slice(start?: number, end?: number): Array<T>;
```
Comment (DanielRosenwasser) [inline realm/index.d.ts]: The linter is complaining about this being `Array<T>` instead of `T[]`.

File: realm/index.d.ts [hunk 5]
```diff
@@ -150,65 +150,65 @@ declare namespace Realm {
...
 
         /**
          * @param  {(object:any,index?:any,collection?:any)=>void} callback
          * @param  {any} thisArg?
          * @returns Object|void
          */
-        find(callback: (object: any, index?: any, collection?: any) => void, thisArg?: any): Object | void;
+        find(callback: (object: any, index?: any, collection?: any) => void, thisArg?: any): (T) | void;
```
Comment (DanielRosenwasser) [inline realm/index.d.ts]: Just `T | null | undefined`

File: realm/index.d.ts [hunk 6]
```diff
@@ -226,43 +226,43 @@ declare namespace Realm {
     * List
     * @see { @link https://realm.io/docs/react-native/latest/api/Realm.List.html }
     */
-    export interface List extends Collection {
+    export interface List<T> extends Collection<T> {
         /**
          * @returns Object|void
          */
-        pop(): Object | void;
+        pop(): (T) | void;
```
Comment (DanielRosenwasser) [inline realm/index.d.ts]: `T | null | undefined`

File: realm/index.d.ts [hunk 7]
```diff
@@ -226,43 +226,43 @@ declare namespace Realm {
...
          * @returns number
          */
-        push(object: any): number;
+        push(object: T): number;
 
         /**
          * @returns Object|void
          */
-        shift(): Object | void;
+        shift(): (T) | void;
```
Comment (DanielRosenwasser) [inline realm/index.d.ts]: `T | null | undefined`

**Review by DanielRosenwasser — CHANGES_REQUESTED** [review]

File: realm/index.d.ts [hunk 8]
```diff
@@ -152,14 +152,14 @@ declare namespace Realm {
          * @param  {number} end?
          * @returns T[] | Object[]
          */
-        slice(start?: number, end?: number): (T & Object)[];
+        slice(start?: number, end?: number): Array<T>;
```
Comment (Akim95) [inline realm/index.d.ts]: Could you update to 'T[]' instead. and at line 252.

Comment (Akim95) [inline realm/index.d.ts]: Could you update to 'T[]'. and at line 252.

**Review by Akim95 — CHANGES_REQUESTED** [review]

**Comment by rodrigopivi** [comment]
@DanielRosenwasser @Akim95 Thanks for the review, updated.

**Comment by rodrigopivi** [comment]
hey @DanielRosenwasser @Akim95 sorry to nag, but is there any chance of this PR getting merge soon? please let me know if you need any further change. thanks!

**Comment by Akim95** [comment]
Tested and anything works fine for me. You should wait maintainer to merge this PR or otherwise.

**Pull request merged**